### PR TITLE
Port eigen_kdl.h/cpp to ROS2

### PR DESCRIPTION
**Pull request by JafarAbdi**
Follow-up to https://github.com/ros2/geometry2/pull/273

### CODE REVIEW

File: eigen_kdl/package.xml
Comment (clalancette) [inline eigen_kdl/package.xml]: Since this now adds `tf2` conversion messages to this code, I'll suggest we rename this to `tf2_eigen_kdl` (this also fits the pattern of the rest of the packages in this repository).

@tfoote What do you think?

File: eigen_kdl/package.xml
Comment (clalancette) [inline eigen_kdl/package.xml]: Can you please enable the linters on this repository, and fix the linting problems?  You should be able to follow https://github.com/ros2/rcpputils/blob/master/package.xml#L15 and https://github.com/ros2/rcpputils/blob/master/CMakeLists.txt#L55 to enable it.

**Review by clalancette — CHANGES_REQUESTED** [review]

File: eigen_kdl/package.xml
Comment (tfoote) [reply eigen_kdl/package.xml]: Yeah that makes sense.

File: tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h
Comment (clalancette) [inline tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h]: Since this is a C++ header file, can you please rename to `tf2_eigen_kdl.hpp`?

File: tf2_eigen_kdl/src/tf2_eigen_kdl.cpp
Comment (clalancette) [inline tf2_eigen_kdl/src/tf2_eigen_kdl.cpp]: Let's clean this up one way or another.  I'd prefer to just remove the commented out code completely, but if you'd like to add a `TODO`, that's OK too.

File: tf2_eigen_kdl/src/tf2_eigen_kdl.cpp
Comment (clalancette) [inline tf2_eigen_kdl/src/tf2_eigen_kdl.cpp]: Style: please use brackets:

```
    for (unsigned int i = 0; i < 3; ++i) {
      e(i, 3) = k.p[i];
    }
```

Same goes for all of the loops below.

File: tf2_eigen_kdl/src/tf2_eigen_kdl.cpp
Comment (clalancette) [inline tf2_eigen_kdl/src/tf2_eigen_kdl.cpp]: The loops above use `unsigned int`; for consistency, I'll suggest doing the same.  Same for all of the methods below.

File: tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h
Comment (clalancette) [inline tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h]: I'm going to go ahead and say that I'm not a great C++ meta-programmer.  Could you please explain what this does?

File: geometry2/package.xml
Comment (clalancette) [inline geometry2/package.xml]: Nit: please put this in alphabetical order.

**Review by clalancette — CHANGES_REQUESTED** [review]
Overall, looking good!  Thanks for iterating.  I've got a few mostly minor things to fix in here.

File: tf2_eigen_kdl/src/tf2_eigen_kdl.cpp
Comment (JafarAbdi) [reply tf2_eigen_kdl/src/tf2_eigen_kdl.cpp]: I did benchmark both methods, they both take the same amount of time

**Comment by JafarAbdi** [comment]
Fixing `uncrustify` causes `cpplint` to fail and vice-versa

**Comment by clalancette** [comment]
> Fixing `uncrustify` causes `cpplint` to fail and vice-versa

There should be a way to satisfy them both.  We have it enabled on almost all of the code in ROS 2, and most of the time you can make both happy.  In the worst case you can add `// NOLINT` to particular lines to quiet down one of them (like https://github.com/ros2/rviz/blob/b8d9b6be6fa75bfa1eb9ffd7f5b94f46229c0342/rviz_default_plugins/include/rviz_default_plugins/displays/camera/camera_display.hpp#L39)

File: tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h
Comment (JafarAbdi) [reply tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h]: We have a member function template `convert` of a class template `Converter`, the first one is for explicitly specializing the class template and the second one is for explicitly specializing the member function template for the two input types

**Comment by JafarAbdi** [comment]
@clalancette Thanks! they're both happy now, I don't know why `uncrustify` was causing `cpplint` to fail

File: tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h
Comment (clalancette) [reply tf2_eigen_kdl/include/tf2_eigen_kdl/tf2_eigen_kdl.h]: Great, thanks for the explanation.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen_kdl/CMakeLists.txt]: I don't think this is necessary; it seems to run the right tests without this.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen_kdl/CMakeLists.txt]: This should also no longer be necessary; once the header file was renamed to `.hpp`, it should detect this automatically.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen_kdl/CMakeLists.txt]: We typically turn on `-Wpedantic` on our code, so please turn that on as well.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen_kdl/CMakeLists.txt]: Nit: use 2 space indentation like everywhere else in this CMakeLists.txt

**Review by clalancette — CHANGES_REQUESTED** [review]
A few more minor changes, then I'm happy to approve and run CI.  Thanks for iterating so far.

**Comment by clalancette** [comment]
Oh, can you also add tests testing out these new methods?  Thanks.

**Comment by clalancette** [comment]
@JafarAbdi friendly ping

**Comment by JafarAbdi** [comment]
> @JafarAbdi friendly ping

Thank you for the reminder, I'll finish this PR this week, sorry for the delay

**Comment by clalancette** [comment]
> Thank you for the reminder, I'll finish this PR this week, sorry for the delay

No problem at all, take your time.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [reply tf2_eigen_kdl/CMakeLists.txt]: This comment is still open.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [inline tf2_eigen_kdl/CMakeLists.txt]: Nit: alphabetize this list.

File: tf2_eigen_kdl/CMakeLists.txt
Comment (clalancette) [reply tf2_eigen_kdl/CMakeLists.txt]: This comment is still open (also, while you are here, please alphabetize this list).

File: tf2_eigen_kdl/package.xml
Comment (clalancette) [inline tf2_eigen_kdl/package.xml]: Please update this to 0.14.1 now.

File: tf2_eigen_kdl/package.xml
Comment (clalancette) [inline tf2_eigen_kdl/package.xml]:
```suggestion
  <maintainer email="clalancette@openrobotics.org">Chris Lalancette</maintainer>
  <maintainer email="alejandro@openrobotics.org">Alejandro Hernandez Cordero</maintainer>
  <license>BSD</license>
  <author>Stuart Glaser</author>
  <author>Adam Leeper</author>

```

File: tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp
Comment (clalancette) [inline tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp]: Two space indentation (same throughout this file).

File: tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp
Comment (clalancette) [inline tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp]:
```suggestion
    return RUN_ALL_TESTS();
```

File: tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp
Comment (clalancette) [inline tf2_eigen_kdl/test/tf2_eigen_kdl_test.cpp]: Add a trailing newline here.

**Review by clalancette — CHANGES_REQUESTED** [review]
Thanks for adding tests!  This is getting closer.  A few more things inline to fix.

**Comment by JafarAbdi** [comment]
Thank you @clalancette for the review! please let me know if there's something I should address

**Review by clalancette — APPROVED** [review]
@JafarAbdi FYI, I did a bit more work here.  In particular:

1.  I rebased onto the latest.
1.  I reenabled the linters, and fixed minor problems pointed out by them.
1.  I updated to modern ament cmake exports.

With all of that in place, I'm now happy with this, so I'm approving.  I'll run CI, but since I did some work here I'd rather get a second review from @ahcorde before merging.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12800/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7759/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10514/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12747/)

**Comment by clalancette** [comment]
Another CI try with Windows fixed:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12801/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7760/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10515/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12748/)

**Review by ahcorde — APPROVED** [review]
LGTM, thank you for the contribution ;)

**Comment by clalancette** [comment]
Approved, and CI is green now, so merging.  Thanks for the contribution.

**Comment by cottsay** [comment]
I believe this PR has regressed builds for Cyclone and Connext:
http://build.ros2.org/view/Rci/job/Rci__nightly-connext_ubuntu_focal_amd64/114/console#console-section-14
http://build.ros2.org/view/Rci/job/Rci__nightly-cyclonedds_ubuntu_focal_amd64/116/console#console-section-12

@JafarAbdi, please take a look ASAP, or we'll need to roll back this change.

**Comment by JafarAbdi** [comment]
@cottsay This PR https://github.com/ros2/geometry2/pull/333 should fix it, is there a way to rerun the build for Cyclone and Connext .?

**Comment by cottsay** [comment]
Thanks for opening a PR so quickly. Hopefully that change will resolve the issue. I've added a comment there with CI that should show if the problem is resolved.

**Pull request merged**
